sklearn/metrics/_regression.py

```diff
@@ -1177,8 +1197,34 @@ def r2_score(
     >>> r2_score(y_true, y_pred, force_finite=False)
     -inf
     """
-    y_type, y_true, y_pred, multioutput = _check_reg_targets(
-        y_true, y_pred, multioutput
+    input_arrays = [y_true, y_pred]
+    if sample_weight is not None:
+        input_arrays.append(sample_weight)
+
+    multioutput_is_array = multioutput is not None and not isinstance(multioutput, str)
+    if multioutput_is_array:
+        input_arrays.append(multioutput)
+
+    xp, is_array_api_compliant = get_namespace(*input_arrays)
+    input_xp = xp
+    device_ = device(*input_arrays)
+
+    if not _supports_dtype(xp, device_, "float64"):
+        y_true = _convert_to_numpy(y_true, xp)
+        y_pred = _convert_to_numpy(y_pred, xp)
```

Comment:
why do we have all this dtype/device specific code in a computation method? It seems to me this doesn't really scale. If we need this kind of code, it should be somewhere used by all other methods. I worry about maintainability of this code.

Reply:
Historically (prior to this PR), `r2_score` always casts its input data to `float64`, even if all inputs are `float32`. We don't know if it's really needed but changing that behavior. For some devices, this is not possible, because they do not support float64 (e.g. only support float32). We don't want to make this method behave numerically differently implicitly based on the device.

The silent conversion to numpy/CPU when the device used to store `y_pred` cannot perform `float64` operations is valid in the case of `r2_score` because we do not expect this to be a performance concern to convert to numpy (computing r2_score is rarely the performance bottleneck of an ML workflow).

This is not the case in general. For estimators code, if the estimator requires upcasts to precision levels not supported by the device, we would rather fail with an explicit error message telling the user that this estimator requires a precision level not supported by the device.

But I think we need to make this explicit in an inline comment here.

Reply:
> We don't want to make this method behave numerically differently implicitly based on the device.

I don't think that's reasonable. I haven't seen people doing loads on random GPUs/devices and expecting the same results. If anything, a part of the experiment is to check numerical stability of the methods on devices for them.

Also, we don't have to be worrying about backward compat since array API is enabled when users enable it via a flag. So that shouldn't be a concern.

Reply:
Let's consider these configurations:

- A: model fit (and evaluated) on fload32 data with numpy
- B: model fit (and evaluated) on fload32 data with torch and device "cuda"
- C: model fit (and evaluated) on fload32 data with torch and device "xps" (macOS GPUs)

Model A and model B would get the same numerical stability because cuda supports upcast to `float64`.

Model `C` would give potentially numerical unstable (invalid) results silently, because it cannot support a `float64` that is only required in the `r2_score` computation (which is not the performance critical step of the workflow).

To me, that would be very confusing.

Reply:
when people deal with different hardware, it will be confusing. AFAIK some hardware do BF16 and some do FP16, and those behave quite differently from one another. We, as scikit-learn, cannot afford to get into that game.

Here, I think what we can do is a "FIX" to keep the dtype on float32 if the input is so, like we do in so many other places. Otherwise, we shouldn't care about the hardware. The whole point of "array API dispatch" is the "dispatch", we don't care what happens in those non-numpy libraries and hardware. The user needs to know their hardware. We will have no guarantee on achieving the same results when somebody uses torch arrays instead of numpy arrays.

Reply:
> Here, I think what we can do is a "FIX" to keep the dtype on float32 if the input is so, like we do in so many other places.

So you mean introducing a behavior change even for numpy on CPU? Or do that change only when non-numpy array api is used? Or only array api is used in conjunction with a device that does not support float64?

We can at least try to remove the upcast and see if existing scikit-learn tests break. If this is the case we would have a concrete example of why float64 upcast is needed for this code.

If not, we cannot conclude that it's not needed though (absence of proof / proof of absence...).

> We will have no guarantee on achieving the same results when somebody uses torch arrays instead of numpy arrays.

In this particular case, A (numpy) and B (torch cuda) are consistent, but B (torch cuda) and C (torch mps) would not behave the same?

Reply:
Alternatively we can leave `r2_score` aside for now and always perform the computation with numpy in all cases, since it's quite unlikely to be performance critical.

This way we guarantee homogeneous behavior.

Reply:
> We can at least try to remove the upcast and see if existing scikit-learn tests break. If this is the case we would have a concrete example of why float64 upcast is needed for this code.

Yes, that's what I mean. Keep computation on input dtype, and see if we need to upcast. And even if we want to upcast unless we can't, it should be in `check_array`, not here.

> but B (torch cuda) and C (torch mps) would not behave the same, which I find very confusing for users.

Either users, or us, can report it upstream, but we shouldn't be shielding users from these issues if they're asking us to dispatch computation to them.

> Alternatively we can leave r2_score aside for now and always perform the computation with numpy in all cases, since it's quite unlikely to be performance critical.

I wouldn't mind.

> This way we guarantee homogeneous behavior.

I don't think we can / should strive for that with array API.

Reply:
> Either users, or us, can report it upstream, but we shouldn't be shielding users from these issues if they're asking us to dispatch computation to them.

What do you mean report upstream? Some devices do not support `float64` arithmetic because they either do not have the hardware support for it (e.g. some integrated Intel GPUs) or they do support such operations in their system APIs (as is the case for [Apple Metal Programming Shaders](https://discuss.pytorch.org/t/code-didnt-speed-up-as-expected-when-using-mps/152016/6)). It's not a bug, it's a known design limitation of such hardware or their system runtime.

If the users give us `float32`-dtyped input on device A, but some intermediate, non-performance critical computation (e.g. computing the r2 score in a CV loop), is not supported by device A (while the performance critical workload such as fitting the model works fast in float32 only on device A).

> it should be in check_array, not here.

`check_array` works with one array at a time. Here the decision to move to CPU should be coordinated between 3 arrays, one of which is optional (`sample_weight`).

Reply:
I did some git archeology to find when the float64 upcast was introduced and it dates back to this PR: #2158. Unfortunately it does not come with a non regression test.

What I find weird with this code is that the sum is performed on positive quantities (assuming weights are positive), so I would not expect [catastrophic cancellations](https://en.wikipedia.org/wiki/Catastrophic_cancellation) in that case.